### ISSUE TITLE
`status` reports typechange of prev subds record as "clean"

### ISSUE THREAD

**Issue by mih** [datalad/datalad#6791]
Discovered when investigating  #5418

Demo

```sh
datalad create ds
cd ds
datalad create -d . subds
rm -rf subds
touch subds
```

at this point `git-status` reports

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        typechange: subds
```

but `datalad-status`

```
{
  "action": "status",
  "gitshasum": "deb687a2fa360d4df134dae95ac552d5c59fd994",
  "parentds": "/tmp/ds",
  "path": "/tmp/ds/subds",
  "prev_gitshasum": "deb687a2fa360d4df134dae95ac552d5c59fd994",
  "refds": "/tmp/ds",
  "state": "clean",
  "status": "ok",
  "type": "dataset"
}
```

**Comment by mih** [comment]
I found that in `GitRepo._diffstatus_get_state_props()` (the status decision-maker) such a typechange looks like this

```
from_state: {'gitshasum': '1271e2ab506ccadeb6072817cc748f68fa626aa1', 'type': 'dataset'}
to_state: {'gitshasum': '1271e2ab506ccadeb6072817cc748f68fa626aa1', 'type': 'dataset'}
modified_in_worktree: True
```

So it seems possible to detect a `typechange` here, possible as easy as including a further conditional

```
elif to_sha == from_sha and modified_in_worktree:
```

**Comment by mih** [comment]
Now, that is not the solution. A simple worktree modification without a typechange looks exactly the same.

It may just be that `GitRepo.get_content_info()` needs to perform an additional `git diff-files` call in order to be able to report types reliably.

**Comment by mih** [comment]
> It may just be that GitRepo.get_content_info() needs to perform an additional git diff-files call in order to be able to report types reliably. Or, likely much cheaper, the worktree modification detection in GitRepo.diffstatus() could switch from ls-files to diff-files.

https://github.com/datalad/datalad/pull/6797 is an attempt to make this happen (switch from `ls-files` to `diff-files`), but it turns out for adjusted branches, it opens a new dimension for corner cases.

In the following scenario `file1` is an empty, annexed file on an adjusted branch:

```sh
% git diff -- file1
diff --git a/file1 b/file1
--- a/file1
+++ b/file1
@@ -1 +0,0 @@
-/annex/objects/MD5E-s0--d41d8cd98f00b204e9800998ecf8427e

% git ls-files --modified -- file1
<no output>

% git diff-files -- file1                                                                              
:100755 100755 7bf9cf177e0559979d401f387afb2cb481c55796 0000000000000000000000000000000000000000 M      file1

% git status -- file1             
On branch adjusted/dl-test-branch(unlocked)
nothing to commit, working tree clean
```

**Comment by yarikoptic** [comment]
Hmm, I wonder if git annex smudge filters might be a culprit, what is you disable them?

**Comment by mih** [comment]
yes, the smudge filters are very likely the culprit, but disabling them is not the solution. The correct outcome is `unmodified`, which can only be reached with the filters on. But diff-files does not have an option to enable them, hence I need to run both (`ls-files` and `diff-files`) to discover modifications and type changes.
And given that realization, this is not just for adjusted-mode. Anyone with smudge filters would run into this.

So the rule for worktree modification detection would be:

if `diff-files` reports a type-change -> `modified`, otherwise go with what `ls-files` reports.

**Comment by mih** [comment]
https://github.com/datalad/datalad-next/pull/589 now brings an implementation that is a basis for an alternative `status` implementation that properly supports such typechange scenarios.
